data_quality/generators/ckan.py

```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+from __future__ import division
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import csv
+import requests
+from os import path
+from data_quality import compat
+from .base import BaseGenerator
+
+class CkanGenerator(BaseGenerator):
+    """This class generates a csv database from a CKAN instance located at the given url"""
+
+    def __init__(self, url=None):
+        """Create an instance if the source url is given.
+
+        Args:
+            url: the base url for the CKAN instance
+        """
+
+        super(CkanGenerator, self).__init__(url)
+
+    def generate_sources(self, sources_filepath, file_types=['csv','excel']):
+        """Generates sources_file from the url"""
+
+        fieldnames = ['id', 'publisher_id', 'title', 'data', 'format', 'period_id']
+        file_types = [ftype.lower() for ftype in file_types]
+        results = self.get_sources()
+        sources = []
+        for result in results:
+            sources += self.extract_sources(result, file_types)
+
+        with compat.UnicodeDictWriter(sources_filepath, fieldnames,
+                                      quoting=csv.QUOTE_MINIMAL) as sfile:
+            sfile.writeheader()
+            for source in sources:
+                sfile.writerow(source)
+
+    def get_sources(self):
+        """Get all sources from CKAN API as a list"""
+
+        extension = 'api/3/action/package_search'
+        full_url = compat.urljoin(self.base_url, extension)
+        response = requests.get(full_url)
+        response.raise_for_status()
+        data = response.json()
+        count = data['result']['count']
+        all_data = []
+        for start in range(0, count, 500):
+            payload = {'rows': 500, 'start': start}
+            response = requests.get(full_url, params=payload)
+            data = response.json()
+            all_data += data['result']['results']
+        return all_data
+
+    def extract_sources(self, datum, file_types):
+        """Extract all sources for one result"""
+
+        resources = []
+        for resource in datum.get('resources', {}):
+            new_resource = {}
+            new_resource['data'] = resource.get('url')
+            ext = path.splitext(new_resource['data'])[1][1:].lower()
+            new_resource['format'] = 'excel' if ext in ['xls', 'xlsx'] else ext
+            file_types = ['excel' if ext in ['xls', 'xlsx'] else ext for ext in file_types]
+            file_types.append('')
+            if new_resource['format'] in file_types:
+                publisher = datum.get('organization', {})
+                new_resource['publisher_id'] = publisher.get('name')
+                new_resource['id'] = resource.get('id')
+                new_resource['data'] = resource.get('url')
+                new_resource['period_id'] = resource.get('last_modified',
```

Comment:
I'm not sure of what timestamp exactly do you need here, but if you really need to know the timestamp the resource **metadata** was modified you can use the `revision_id` on the resource to get its details with `revision_show`:

http://demo.ckan.org/api/action/revision_show?id=2e2e7a3c-68aa-4b8c-b86b-6345581ca10e

This will have a `timestamp`. Note that this could mean changes in any field of the resource, url, title, etc...